### DISCUSSION TITLE
Use atomic inside YARP

### DISCUSSION THREAD

**Discussion by Nicogene** [robotology/yarp#1754] [Ideas]
During the refactoring done for the release of YARP 3.0.0 @drdanz and me found that `yarp` is plenty of code like this

```
class Foo 
{
   int count;
   std::mutex;
   void doStuff()
   {
       mutex.lock();
       count++;
       mutex.unlock();
   }
};

```

that should become

```
class Foo 
{
   std::atomic<int> count;
   void doStuff()
   {
       count++;
   }
};

```

`atomic` will free us from a lot of mutexes/semaphores

**Comment by barbalberto** [comment]
For **boolean** flags (bool only) there is also `std::atomic_flag` which is atomic and **lock-free**

From https://en.cppreference.com/w/cpp/atomic/atomic_flag
> std::atomic_flag is an atomic boolean type. Unlike all specializations of std::atomic, it is guaranteed to be lock-free.
